File: acme-admin/src/main/java/com/acme/admin/controller/security/UserAdminController.java
package com.acme.admin.controller.security;

import com.acme.admin.dto.security.UserDtos.*;
import com.acme.admin.service.security.UserService;
import jakarta.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.*;

@RestController
@RequestMapping("/api/admin/users")
public class UserAdminController {
  private final UserService users;

  public UserAdminController(UserService users) {
    this.users = users;
  }

  @GetMapping
  @PreAuthorize("hasAuthority('users.read')")
  public List<UserDto> list() {
    return users.list();
  }

  @PostMapping
  @PreAuthorize("hasAuthority('users.write')")
  public UserDto create(@Valid @RequestBody CreateUserRequest req) {
    return users.create(req);
  }

  @PutMapping("/{id}")
  @PreAuthorize("hasAuthority('users.write')")
  public UserDto update(@PathVariable UUID id, @Valid @RequestBody UpdateUserRequest req) {
    return users.update(id, req);
  }
}

File: acme-admin-dotnet/Controllers/UserAdminController.cs
using Acme.Admin.Api.DTO;
using Acme.Admin.Api.Security;
using Acme.Admin.Api.Services;
using Acme.Admin.Api.Validation;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Acme.Admin.Api.Controllers;

[ApiController]
[Route("api/admin/users")]
public sealed class UserAdminController(UserService service) : ControllerBase
{
    [HttpGet]
    [Authorize(Policy = PermissionPolicies.UsersRead)]
    public async Task<ActionResult<PagedResult<UserDtos.UserDto>>> List(
        [FromQuery] int page = Paging.DefaultPage,
        [FromQuery] int size = Paging.DefaultSize,
        CancellationToken ct = default)
    {
        var (p, s) = Paging.Normalize(page, size);
        return Ok(await service.ListAsync(p, s, ct));
    }

    [HttpPost]
    [Authorize(Policy = PermissionPolicies.UsersWrite)]
    public async Task<ActionResult<UserDtos.UserDto>> Create([FromBody] UserDtos.CreateUserRequest req, CancellationToken ct)
    {
        return Ok(await service.CreateAsync(req, ct));
    }

    [HttpPut("{id:guid}")]
    [Authorize(Policy = PermissionPolicies.UsersWrite)]
    public async Task<ActionResult<UserDtos.UserDto>> Update([FromRoute] Guid id, [FromBody] UserDtos.UpdateUserRequest req, CancellationToken ct)
    {
        return Ok(await service.UpdateAsync(id, req, ct));
    }
}
